File: special_hit.py
from globals import *

__all__ = ['special_hit', 'cough_up', 'orc_gold', 'check_orc', 'check_xeroc',
           'hiding_xeroc', 'm_confuse', 'flame_broil']

def special_hit(monster):
    k = monster.ichar
    if k == 'A':
        rust(monster)
    elif k == 'F':
        g.being_held = 1
    elif k == 'I':
        freeze(monster)
    elif k == 'L':
        steal_gold(monster)
    elif k == 'N':
        steal_item(monster)
    elif k == 'R':
        sting(monster)
    elif k == 'V':
        drain_life()
    elif k == 'W':
        drain_level()

def rust(monster):
    if not rogue.armor or get_armor_class(rogue.armor) <= 1 or rogue.armor.which_kind == LEATHER:
        return
    if rogue.armor.is_protected:
        if not monster.identified:
            message("the rust vanishes instantly", 0)
            monster.identified = 1
    else:
        rogue.armor.damage_enchantment -= 1
        message("your armor weakens", 0)
        print_stats()

def freeze(monster):
    if rand_percent(12): return
    
    freeze_percent = 99
    freeze_percent -= rogue.strength_current + rogue.strength_current / 2
    freeze_percent -= rogue.exp * 4
    freeze_percent -= get_armor_class(rogue.armor) * 5
    freeze_percent -= rogue.hp_max / 3
    
    if freeze_percent > 10:
        monster.identified = 1
        message("you are frozen", 1)
        
        n = get_rand(5, 9)
        for i in range(n):
            move_monsters()
        if rand_percent(freeze_percent):
            for i in range(50):
                move_monsters()
            killed_by(None, HYPOTHERMIA)
        message("you can move again", 1)
        monster.identified = 0

def steal_gold(monster):
    if rand_percent(15): return
    
    if rogue.gold > 50:
        amount = get_rand(8, 15) if rogue.gold > 1000 else get_rand(2, 5)
        amount = rogue.gold / amount
    else:
        amount = rogue.gold / 2
    amount += (get_rand(0, 2) - 1) * (rogue.exp + g.current_level)
    
    if amount <= 0 and rogue.gold > 0:
        amount = rogue.gold
    
    if amount > 0:
        rogue.gold -= amount
        message("your purse feels lighter", 0)
        print_stats()
        
    disappear(monster)

def steal_item(monster):
    if rand_percent(15): return
    
    has_something = 0
    obj = rogue.pack.next_object
    while obj:
        if obj != rogue.armor and obj != rogue.weapon:
            has_something = 1
            break
        obj = obj.next_object
    if has_something:
        n = get_rand(0, MAX_PACK_COUNT)
        obj = rogue.pack.next_object
        
        for i in range(n + 1):
            obj = obj.next_object
            while not obj or obj == rogue.armor or obj == rogue.weapon:
                if not obj:
                    obj = rogue.pack.next_object
                else:
                    obj = obj.next_object
        message("she stole " + get_description(obj), 0)
        
        if obj.what_is == AMULET:
            g.has_amulet = 0
        vanish(obj, 0)
    disappear(monster)

def disappear(monster):
    row = monster.row
    col = monster.col
    
    remove_mask(row, col, MONSTER)
    if can_see(row, col):
        mvaddch(row, col, get_room_char(screen[row][col], row, col))
    remove_from_pack(monster, g.level_monsters)

def cough_up(monster):
    if g.current_level < g.max_level: return
    
    if monster.ichar == 'L':
        obj = get_an_object()
        obj.what_is = GOLD
        obj.quantity = get_rand(9, 599)
    else:
        if rand_percent(monster.which_kind):
            obj = get_rand_object()
        else:
            return
    
    row = monster.row
    col = monster.col
    
    for n in range(6):
        for i in range(-n, n + 1):
            if try_to_cough(row + n, col + i, obj):
                return
            if try_to_cough(row - n, col + i, obj):
                return
        for i in range(-n, n + 1):
            if try_to_cough(row + i, col - n, obj):
                return
            if try_to_cough(row + i, col + n, obj):
                return

def try_to_cough(row, col, obj):
    if row < MIN_ROW or row > LINES - 2 or col < 0 or col > COLS - 1:
        return 0
    if not screen[row][col] & IS_OBJECT and not screen[row][col] & MONSTER and screen[row][col] & (TUNNEL | FLOOR | DOOR):
        put_object_at(obj, row, col)
        mvaddch(row, col, get_room_char(screen[row][col], row, col))
        refresh()
        return 1
    return 0

def orc_gold(monster):
    if monster.identified:
        return 0
    rn = get_room_number(monster.row, monster.col)
    if rn < 0:
        return 0
    r = rooms[rn]
    for i in range(r.top_row + 1, r.bottom_row):
        for j in range(r.left_col + 1, r.right_col):
            if screen[i][j] & GOLD and not screen[i][j] & MONSTER:
                monster.m_flags |= CAN_GO
                s = monster_can_go(monster, i, j)
                monster.m_flags &= ~CAN_GO
                if s:
                    move_monster_to(monster, i, j)
                    monster.m_flags |= IS_ASLEEP
                    monster.m_flags &= ~WAKENS
                    monster.identified = 1
                    return 1
                monster.identified = 1
                monster.m_flags |= CAN_GO
                mv_monster(monster, i, j)
                monster.m_flags &= ~CAN_GO
                monster.identified = 0
                return 1
    return 0
    
def check_orc(monster):
    if monster.ichar == 'O':
        monster.identified = 1

def check_xeroc(monster):
    if monster.ichar == 'X' and monster.identified:
        wake_up(monster)
        monster.identified = 0
        mvaddch(monster.row, monster.col, get_room_char(screen[monster.row][monster.col], monster.row, monster.col))
        check_message()
        message("wait, that's a %s!" % monster_name(monster), 1)
        return 1
    return 0

def hiding_xeroc(row, col):
    if g.current_level < XEROC1 or g.current_level > XEROC2 or not screen[row][col] & MONSTER:
        return 0
    
    monster = object_at(g.level_monsters, row, col)
    return monster.ichar == 'X' and monster.identified
    
def sting(monster):
    if rogue.strength_current < 5: return
    
    sting_chance = 35
    ac = get_armor_class(rogue.armor)
    sting_chance += 6 * (6 - ac)
    
    if rogue.exp > 8:
        sting_chance -= 6 * (rogue.exp - 8)
    
    sting_chance = max(min(sting_chance, 100), 1)
    
    if rand_percent(sting_chance):
        message("the %s's bite has weakened you" % monster_name(monster), 0)
        rogue.strength_current -= 1
        print_stats()
    
def drain_level():
    if not rand_percent(20) or rogue.exp < 8:
        return
    
    rogue.exp_points = level_points[rogue.exp - 2] - get_rand(10, 50)
    rogue.exp -= 2
    add_exp(1)
    
def drain_life():
    if not rand_percent(25) or rogue.hp_max <= 30 or rogue.hp_current < 10:
        return
    message("you feel weaker", 0)
    rogue.hp_max -= 1
    rogue.hp_current -= 1
    if rand_percent(50):
        if rogue.strength_current >= 5:
            rogue.strength_current -= 1
            if rand_percent(50):
                rogue.strength_max -= 1
    print_stats()

def m_confuse(monster):
    if monster.identified:
        return 0
    if not can_see(monster.row, monster.col):
        return 0
    if rand_percent(45):
        monster.identified = 1
        return 0
    if rand_percent(55):
        monster.identified = 1
        message("the gaze of the %s has confused you" % monster_name(monster), 1)
        confuse()
        return 1
    return 0

def flame_broil(monster):
    if rand_percent(50):
        return 0
    row, col = monster.row, monster.col
    if not can_see(row, col):
        return 0
    if not rogue_is_around(row, col):
        row, col = get_closer(row, col, rogue.row, rogue.col)
        standout()
        while True:
            mvaddch(row, col, '*')
            refresh()
            row, col = get_closer(row, col, rogue.row, rogue.col)
            if row == rogue.row and col == rogue.col: break
        standend()
        
        row, col = get_closer(monster.row, monster.col, rogue.row, rogue.col)
        while True:
            mvaddch(row, col, get_room_char(screen[row][col], row, col))
            refresh()
            row, col = get_closer(row, col, rogue.row, rogue.col)
            if row == rogue.row and col == rogue.col: break
    monster_hit(monster, "flame")
    return 1
    
def get_closer(row, col, trow, tcol):
    if row < trow:
        row += 1
    elif row > trow:
        row -= 1
    if col < tcol:
        col += 1
    elif col > tcol:
        col -= 1
    return row, col

File: flrogue/lib/rogue/special_hit.dart
import 'globals.dart';
import 'hit.dart';
import 'level.dart';
import 'message.dart';
import 'monster.dart';
import 'object.dart';
import 'pack.dart';
import 'room.dart';
import 'score.dart';
import 'ui.dart';
import 'use.dart';

Future<void> specialHit(GameObject monster) async {
  String k = monster.ichar;
  if (k == 'A') {
    await _rust(monster);
  } else if (k == 'F') {
    beingHeld = true;
  } else if (k == 'I') {
    await _freeze(monster);
  } else if (k == 'L') {
    await _stealGold(monster);
  } else if (k == 'N') {
    await _stealItem(monster);
  } else if (k == 'R') {
    await _sting(monster);
  } else if (k == 'V') {
    await _drainLife();
  } else if (k == 'W') {
    await _drainLevel();
  }
}

Future<void> _rust(GameObject monster) async {
  if (rogue.armor == null ||
      getArmorClass(rogue.armor) <= 1 ||
      rogue.armor!.whichKind == ArmorType.leather.index) {
    return;
  }

  if (rogue.armor!.isProtected != 0) {
    if (monster.identified == 0) {
      await message("the rust vanishes instantly");
      monster.identified = 1;
    }
  } else {
    rogue.armor!.damageEnchantment -= 1;
    await message("your armor weakens");
    printStats();
  }
}

Future<void> _freeze(GameObject monster) async {
  if (!randPercent(12)) return;

  int freezePercent = 99;
  freezePercent -= rogue.strengthCurrent + rogue.strengthCurrent ~/ 2;
  freezePercent -= rogue.exp * 4;
  freezePercent -= getArmorClass(rogue.armor) * 5;
  freezePercent -= rogue.hpMax ~/ 3;

  if (freezePercent > 10) {
    monster.identified = 1;
    await message("you are frozen", true);

    int n = getRand(5, 9);
    for (int i = 0; i < n; i++) {
      await moveMonsters();
    }

    if (randPercent(freezePercent)) {
      for (int i = 0; i < 50; i++) {
        await moveMonsters();
      }
      await killedBy(null, DeathCause.hypothermia);
    }

    await message("you can move again", true);
    monster.identified = 0;
  }
}

Future<void> _stealGold(GameObject monster) async {
  if (!randPercent(15)) return;

  int amount;
  if (rogue.gold > 50) {
    amount = rogue.gold > 1000 ? getRand(8, 15) : getRand(2, 5);
    amount = rogue.gold ~/ amount;
  } else {
    amount = rogue.gold ~/ 2;
  }

  amount += (getRand(0, 2) - 1) * (rogue.exp + currentLevel);

  if (amount <= 0 && rogue.gold > 0) {
    amount = rogue.gold;
  }

  if (amount > 0) {
    rogue.gold -= amount;
    await message("your purse feels lighter");
    printStats();
  }

  _disappear(monster);
}

Future<void> _stealItem(GameObject monster) async {
  if (!randPercent(15)) return;

  int items = 0;
  for (GameObject obj in rogue.pack) {
    if (obj != rogue.armor && obj != rogue.weapon) {
      items++;
    }
  }

  if (items > 0) {
    int n = getRand(0, items);

    GameObject obj = rogue.pack.first;
    for (obj in rogue.pack) {
      if (obj != rogue.armor && obj != rogue.weapon) {
        if (--n < 0) {
          break;
        }
      }
    }

    await message("she stole ${getDescription(obj)}");

    if (obj.whatIs == Cell.amulet) {
      hasAmulet = false;
    }

    await vanish(obj, false);
  }

  _disappear(monster);
}

void _disappear(GameObject monster) {
  int row = monster.row;
  int col = monster.col;

  removeMask(row, col, Cell.monster);
  if (canSee(row, col)) {
    ui.move(row, col);
    ui.write(getRoomChar(screen[row][col], row, col));
  }

  removeFromPack(monster, levelMonsters);
}

void coughUp(GameObject monster) {
  if (currentLevel < maxLevel) return;

  GameObject obj;

  if (monster.ichar == 'L') {
    obj = getAnObject();
    obj.whatIs = Cell.gold;
    obj.quantity = getRand(9, 599);
  } else {
    if (randPercent(monster.whichKind)) {
      obj = getRandObject();
    } else {
      return;
    }
  }

  int row = monster.row;
  int col = monster.col;

  for (int n = 0; n < 6; n++) {
    for (int i = -n; i <= n; i++) {
      if (_tryToCough(row + n, col + i, obj)) {
        return;
      }
      if (_tryToCough(row - n, col + i, obj)) {
        return;
      }
    }

    for (int i = -n; i <= n; i++) {
      if (_tryToCough(row + i, col - n, obj)) {
        return;
      }
      if (_tryToCough(row + i, col + n, obj)) {
        return;
      }
    }
  }
}

bool _tryToCough(int row, int col, GameObject obj) {
  if (row < minRow || row > ui.rows - 2 || col < 0 || col > ui.cols - 1) {
    return false;
  }

  if (!(screen[row][col] & Cell.isObject != 0) &&
      !(screen[row][col] & Cell.monster != 0) &&
      (screen[row][col] & (Cell.tunnel | Cell.floor | Cell.door) != 0)) {
    putObjectAt(obj, row, col);
    ui.move(row, col);
    ui.write(getRoomChar(screen[row][col], row, col));
    ui.refresh();
    return true;
  }

  return false;
}

Future<bool> orcGold(GameObject monster) async {
  if (monster.identified != 0) {
    return false;
  }

  int rn = getRoomNumber(monster.row, monster.col);
  if (rn < 0) {
    return false;
  }

  Room r = rooms[rn];
  for (int i = r.topRow + 1; i < r.bottomRow; i++) {
    for (int j = r.leftCol + 1; j < r.rightCol; j++) {
      if (screen[i][j] & Cell.gold != 0 &&
          !(screen[i][j] & Cell.monster != 0)) {
        monster.flagsAdd(MonsterFlags.canGo);
        bool s = monsterCanGo(monster, i, j);
        monster.flagsRemove(MonsterFlags.canGo);

        if (s) {
          moveMonsterTo(monster, i, j);
          monster.flagsAdd(MonsterFlags.isAsleep);
          monster.flagsRemove(MonsterFlags.wakens);
          monster.identified = 1;
          return true;
        }

        monster.identified = 1;
        monster.flagsAdd(MonsterFlags.canGo);
        await mvMonster(monster, i, j);
        monster.flagsRemove(MonsterFlags.canGo);
        monster.identified = 0;
        return true;
      }
    }
  }

  return false;
}

void checkOrc(GameObject monster) {
  if (monster.ichar == 'O') {
    monster.identified = 1;
  }
}

Future<bool> checkXeroc(GameObject monster) async {
  if (monster.ichar == 'X' && monster.identified != 0) {
    wakeUp(monster);
    monster.identified = 0;
    ui.move(monster.row, monster.col);
    ui.write(
      getRoomChar(screen[monster.row][monster.col], monster.row, monster.col),
    );
    checkMessage();
    await message("wait, that's a ${monsterName(monster)}!", true);
    return true;
  }
  return false;
}

bool hidingXeroc(int row, int col) {
  if (currentLevel < xeroc1 ||
      currentLevel > xeroc2 ||
      !(screen[row][col] & Cell.monster != 0)) {
    return false;
  }

  GameObject monster = objectAt(levelMonsters, row, col)!;
  return monster.ichar == 'X' && monster.identified != 0;
}

Future<void> _sting(GameObject monster) async {
  if (rogue.strengthCurrent < 5) return;

  int stingChance = 35;
  int ac = getArmorClass(rogue.armor);
  stingChance += 6 * (6 - ac);

  if (rogue.exp > 8) {
    stingChance -= 6 * (rogue.exp - 8);
  }

  stingChance = stingChance > 100 ? 100 : stingChance;
  stingChance = stingChance < 1 ? 1 : stingChance;

  if (randPercent(stingChance)) {
    await message("the ${monsterName(monster)}'s bite has weakened you");
    rogue.strengthCurrent -= 1;
    printStats();
  }
}

Future<void> _drainLevel() async {
  if (!randPercent(20) || rogue.exp < 8) {
    return;
  }

  rogue.expPoints = levelPoints[rogue.exp - 2] - getRand(10, 50);
  rogue.exp -= 2;
  await addExp(1);
}

Future<void> _drainLife() async {
  if (!randPercent(25) || rogue.hpMax <= 30 || rogue.hpCurrent < 10) {
    return;
  }

  await message("you feel weaker");
  rogue.hpMax -= 1;
  rogue.hpCurrent -= 1;

  if (randPercent(50)) {
    if (rogue.strengthCurrent >= 5) {
      rogue.strengthCurrent -= 1;
      if (randPercent(50)) {
        rogue.strengthMax -= 1;
      }
    }
  }

  printStats();
}

Future<bool> mConfuse(GameObject monster) async {
  if (monster.identified != 0) {
    return false;
  }

  if (!canSee(monster.row, monster.col)) {
    return false;
  }

  if (randPercent(45)) {
    monster.identified = 1;
    return false;
  }

  if (randPercent(55)) {
    monster.identified = 1;
    await message(
      "the gaze of the ${monsterName(monster)} has confused you",
      true,
    );
    confuse();
    return true;
  }

  return false;
}

Future<bool> flameBroil(GameObject monster) async {
  if (!randPercent(50)) {
    return false;
  }

  int row = monster.row, col = monster.col;
  if (!canSee(row, col)) {
    return false;
  }

  if (!rogueIsAround(row, col)) {
    (row, col) = _getCloser(row, col, rogue.row, rogue.col);

    ui.move(row, col);
    ui.write('*', inverse: true);

    while (row != rogue.row || col != rogue.col) {
      (row, col) = _getCloser(row, col, rogue.row, rogue.col);

      if (row == rogue.row && col == rogue.col) break;

      ui.move(row, col);
      ui.write('*', inverse: true);
      ui.refresh();
    }

    row = monster.row;
    col = monster.col;
    (row, col) = _getCloser(row, col, rogue.row, rogue.col);

    while (row != rogue.row || col != rogue.col) {
      ui.move(row, col);
      ui.write(getRoomChar(screen[row][col], row, col));
      ui.refresh();

      (row, col) = _getCloser(row, col, rogue.row, rogue.col);

      if (row == rogue.row && col == rogue.col) break;
    }
  }

  await monsterHit(monster, "flame");
  return true;
}

(int, int) _getCloser(int row, int col, int trow, int tcol) {
  if (row < trow) {
    row += 1;
  } else if (row > trow) {
    row -= 1;
  }

  if (col < tcol) {
    col += 1;
  } else if (col > tcol) {
    col -= 1;
  }

  return (row, col);
}
